[Structural] Eigensolver improve Dirichlet BC handling

philbucher opened the pull request:
@emiroglu had some issues with the dirichlet boundary conditions in the EigenSolver with the BlockB&S
We noticed that the application of them is a bit odd (see long discussion in #279)
I spoke already with @AndreasWinterstein and we think that it was done like this due to the awkward interface of FEAST (which in the meantime is practically no longer used afaik).

We will discuss it at the chair together with @qaumann who was also involved in #279 on Thursday and then we can discuss here further

Also a fix is added when using the EliminationB&S

@emiroglu those are the cleaned changes from this afternoon, you can test it with this branch

philbucher committed:
fixes for BC application

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:
```diff
@@ -687,12 +690,15 @@ class EigensolverStrategy
                 {
                     auto itDof = std::begin(NodeDofs) + j;
-                    rNodeEigenvectors(i,j) = rEigenvectors(i,(*itDof)->EquationId());
+                    if ((*itDof)->IsFree()) {
```
this is required for the EliminationB&S because the solution (`rEigenvectors`) does not contain the fixed dofs
Guess none tried this before :)

qaumann replied:
now this is, where this segmentation fault came from...

RiccardoRossi replied:
MINOR: this loop should be made openmp parallel...

philbucher replied:
If we want to do this then we have to refactor the [entire loop](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp#L671-L697) bcs of the allocation in parallel of the DataValueContainer [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp#L675)
I can do it but only after we converged on how to proceed with this

emiroglu replied:
there is another problem in this line:
For cases with mixed elements (e.g. trusses & beams) on nodes that belong only to trusses we also have the `ROTATION` dofs. These dofs are not in the `DofSet` in the B&S, but they are still there. Meaning that if we loop the dofs (which we are doing here) then we also get those "inactive" dofs. They return the default equation Id which is 0 and leads to wrong results.
I don't know to differentiate the "active" from the "inactive" dofs. We thought of setting an equation Id of -1 but this does not work bcs it is a `std::size_t`

Maybe someone has an idea?

loumalouomega commented:
Question ?, it is solved if using the B&S with improved condition number?

El mié., 19 feb. 2020 1:43, Philipp Bucher <notifications@github.com>
escribió:

> *@philbucher* commented on this pull request.
> ------------------------------
>
> In
> applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
> <https://github.com/KratosMultiphysics/Kratos/pull/6434#discussion_r381018203>
> :
>
> >                  for (std::size_t j = 0; j < NumNodeDofs; j++)
>                  {
>                      auto itDof = std::begin(NodeDofs) + j;
> -                    rNodeEigenvectors(i,j) = rEigenvectors(i,(*itDof)->EquationId());
> +                    if ((*itDof)->IsFree()) {
>
> this is required for the EliminationB&S because the solution (
> rEigenvectors) does not contain the fixed dofs
> Guess none tried this before :)
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6434?email_source=notifications&email_token=AEYQZAAPQ3FJ5SUS7SJ5CZLRDR6BHA5CNFSM4KXPI5XKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCWAPRZI#pullrequestreview-360773861>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYQZAAYQRBPSZX25EKRT6TRDR6BHANCNFSM4KXPI5XA>
> .
>

loumalouomega commented:
I cannot check the PR right now...

El mié., 19 feb. 2020 8:30, Vicente Mataix Ferrándiz <
vicente.mataix@polytechnique.edu> escribió:

> Question ?, it is solved if using the B&S with improved condition number?
>
> El mié., 19 feb. 2020 1:43, Philipp Bucher <notifications@github.com>
> escribió:
>
>> *@philbucher* commented on this pull request.
>> ------------------------------
>>
>> In
>> applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
>> <https://github.com/KratosMultiphysics/Kratos/pull/6434#discussion_r381018203>
>> :
>>
>> >                  for (std::size_t j = 0; j < NumNodeDofs; j++)
>>                  {
>>                      auto itDof = std::begin(NodeDofs) + j;
>> -                    rNodeEigenvectors(i,j) = rEigenvectors(i,(*itDof)->EquationId());
>> +                    if ((*itDof)->IsFree()) {
>>
>> this is required for the EliminationB&S because the solution (
>> rEigenvectors) does not contain the fixed dofs
>> Guess none tried this before :)
>>
>> —
>> You are receiving this because your review was requested.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/KratosMultiphysics/Kratos/pull/6434?email_source=notifications&email_token=AEYQZAAPQ3FJ5SUS7SJ5CZLRDR6BHA5CNFSM4KXPI5XKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCWAPRZI#pullrequestreview-360773861>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEYQZAAYQRBPSZX25EKRT6TRDR6BHANCNFSM4KXPI5XA>
>> .
>>
>

qaumann commented:
I looked a bit into it and the problem might be, that also the zero diagonals of both matrices are set to 1. This creates spurious eigenvalues related to the value of the diagonal entry. If the diagonal of the second matrix (here the mass amtrix) is set to zero, the eigenvalues should be correct. In the FEAST case, the factor with which the diagonals are scaled should be 0 for the mass matrix and any value for the stiffness matrix.

Let's see, how the solvers handle the zeros on the diagonal, but at a first glance, the FEAST from `ExternalSolversApplication` should be fine with it.

Edit: what is done now in the `EigensolverStrategy` to overcome the problem with positive definiteness is to set the diagonal to 1 in the mass matrix and to -1 in the stiffness matrix. This results in all spurious eigenvalues being -1, which are then excluded from the eigenvalue search using FEAST's search contour. However, this is only a valid approach, if it is known that no eigenvalues of interest are negative.

(some reference: https://scicomp.stackexchange.com/questions/31140/dirichlet-boundary-conditions-in-generalized-eigenvalue-problem)

loumalouomega commented:
That's why I asked to use the improved version of the B&S.

El mié., 19 feb. 2020 9:20, qaumann <notifications@github.com> escribió:

> I looked a bit into it and the problem might be, that also the zero
> diagonals of both matrices are set to 1. This creates spurious eigenvalues
> related to the value of the diagonal entry. If the diagonal of the second
> matrix (here the mass amtrix) is set to zero, the eigenvalues should be
> correct. In the FEAST case, the factor with which the diagonals are scaled
> should be 0 for the mass matrix and any value for the stiffness matrix.
>
> Let's see, how the solvers handle the zeros on the diagonal, but at a
> first glance, the FEAST from ExternalSolversApplication should be fine
> with it.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6434?email_source=notifications&email_token=AEYQZAED3H7H3B7O4SBJ4G3RDTTVJA5CNFSM4KXPI5XKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMGZO6Y#issuecomment-588093307>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYQZAFVWX7UZVSSPSCS3B3RDTTVJANCNFSM4KXPI5XA>
> .
>

qaumann commented:
@loumalouomega where can I find it?

loumalouomega commented:
> 
> 
> I looked a bit into it and the problem might be, that also the zero diagonals of both matrices are set to 1. This creates spurious eigenvalues related to the value of the diagonal entry. If the diagonal of the second matrix (here the mass amtrix) is set to zero, the eigenvalues should be correct. In the FEAST case, the factor with which the diagonals are scaled should be 0 for the mass matrix and any value for the stiffness matrix.
> 
> Let's see, how the solvers handle the zeros on the diagonal, but at a first glance, the FEAST from `ExternalSolversApplication` should be fine with it.
> 
> Edit: what is done now in the `EigensolverStrategy` to overcome the problem with positive definiteness is to set the diagonal to 1 in the mass matrix and to -1 in the stiffness matrix. This results in all spurious eigenvalues being -1, which are then excluded from the eigenvalue search using FEAST's search contour. However, this is only a valid approach, if it is known that no eigenvalues of interest are negative.
> 
> (some reference: https://scicomp.stackexchange.com/questions/31140/dirichlet-boundary-conditions-in-generalized-eigenvalue-problem)

See https://github.com/KratosMultiphysics/Kratos/pull/5905, may solve your problem

loumalouomega commented:
> 
> 
> @loumalouomega where can I find it?

https://github.com/KratosMultiphysics/Kratos/pull/5905

loumalouomega commented:
> 
> 
> > @loumalouomega where can I find it?
> 
> #5905

If this solves the problem it means that....
![imagen](https://user-images.githubusercontent.com/19991680/74816950-68b2b580-52fc-11ea-94c9-2718651b475a.png)

qaumann commented:
I think it is definately a related problem, but there will still be spurious eigenvalues related to the fixed dofs. It even makes things a bit worse, as the spurious eigenvalues are not necessarily the lowest or highest. Would be interesting to see though, how the corresponding eigenvectors look like. I will have a look at this.

qaumann commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:
```diff
@@ -402,7 +405,7 @@ class EigensolverStrategy
         if (rModelPart.NumberOfMasterSlaveConstraints() != 0) {
             this->pGetBuilderAndSolver()->ApplyConstraints(pScheme, rModelPart, rMassMatrix, b);
         }
-        this->ApplyDirichletConditions(rMassMatrix, 1.0);
```
Because of the reasons stated in the comments, this change would produce wrong results for a `BlockBuilderAndSolver`

qaumann commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:
```diff
@@ -416,7 +419,7 @@ class EigensolverStrategy
         if (rModelPart.NumberOfMasterSlaveConstraints() != 0) {
             this->pGetBuilderAndSolver()->ApplyConstraints(pScheme, rModelPart, rStiffnessMatrix, b);
         }
-        ApplyDirichletConditions(rStiffnessMatrix,-1.0);
```
Because of the reasons stated in the comments, this change would produce wrong results for a `BlockBuilderAndSolver`

qaumann replied:
Setting the factor to 0 here (and leaving the factor for the mass matrix at 1) will set the spurious eigenvalues to 0, which can be filtered easily. Maybe this is a way to proceed?

emiroglu commented:
I have tried the suggested modifications with different combinations (block/elimination B&S) and I keep getting either wrong eigenvalues or wrong results at some of the constrained nodes for random modeshapes. I will make a list of combinations and corresponding errors tomorrow if the discussion with @philbucher @qaumann does not resolve this. I will also upload the case I am considering.

by the way @qaumann : Giving the factor as 0 for block B&S gives the following error: 
THE MATRIX IS STRUCTURALLY SINGULAR ... ZERO COLUMN AT ...

loumalouomega commented:
I may have an idea, @emiroglu can you try to set to zero the terms of the mass matrix corresponding to the fixed dofs? (this may be the reason)

qaumann commented:
> I have tried the suggested modifications with different combinations (block/elimination B&S) and I keep getting either wrong eigenvalues or wrong results at some of the constrained nodes for random modeshapes. I will make a list of combinations and corresponding errors tomorrow if the discussion with @philbucher @qaumann does not resolve this. I will also upload the case I am considering.
> 
> by the way @qaumann : Giving the factor as 0 for block B&S gives the following error:
> THE MATRIX IS STRUCTURALLY SINGULAR ... ZERO COLUMN AT ...

@emiroglu This happens, if the mass matrix is singular beforehand. Try uncommenting lines 610-624 in eigensolver_strategy.hpp, then it should work.

You also have to exclude 0 in the search interval ("lambda_min" > 0), otherwise the "zero" eigenvalues will still be found.

philbucher committed:
review comments

philbucher committed:
using elim-b&s for eigensolver by default

philbucher committed:
formatting

emiroglu committed:
making it work for both B&S

not the cleanest solution, can be hopefully improved with #5905

emiroglu committed:
revert using elimB&S by default

RiccardoRossi commented:
Hi guys, one word about this:

the proposal in #5905 could be a clean solution about this.
Indeed one could choose a value of the diagonal on dirichlet nodes so that the corresponding eigenvalue is the highest in the system. that would effectively solve the troubles you have with the lowest eigenvalues.

Basically what i am telling is that the diagonal value you choose is nothing else than the eigenvalue corresponding to an eigenvector with a single 1 corresponding to that component.

emiroglu committed:
improvement

emiroglu committed:
expose new constructor

emiroglu committed:
new constructor

philbucher commented:
@RiccardoRossi you are right that #5905 might be a clean solution, however not as it currently is!
The problem is that the scaling factor has to be given in the constructor.
Since we different ones for MassMatrix and StiffnessMatrix it would be required to set this factor in the `ProcessInfo`
@loumalouomega I think you initially had the factor in the ProccessInfo, why did you add it to the constructor? It is less flexibel

loumalouomega commented:
> @loumalouomega I think you initially had the factor in the ProccessInfo, why did you add it to the constructor? It is less flexibel

@RiccardoRossi said that... (in any case we can add a method to the builder and solver to retrieve the value, or assign to a variable again)

philbucher commented:
You commented in the other PR but I cannot find the use of `SCALE_DIAGONAL` :/

The problem with adding another function is that It would have to be added to the baseclass in order to work. Do we want that?

loumalouomega commented:
> 
> 
> You commented in the other PR but I cannot find the use of `SCALE_DIAGONAL` :/
> 
> The problem with adding another function is that It would have to be added to the baseclass in order to work. Do we want that?

Indeed. Just suggesting, can be:

~~~cpp
virtual double GetDiagonalScaleFactor() const
{
    return 1.0;
}
~~~

loumalouomega commented:
> 
> 
> > You commented in the other PR but I cannot find the use of `SCALE_DIAGONAL` :/
> > The problem with adding another function is that It would have to be added to the baseclass in order to work. Do we want that?
> 
> Indeed. Just suggesting, can be:
> 
> ```c++
> virtual double GetDiagonalScaleFactor() const
> {
>     return 1.0;
> }
> ```

Opinion @RiccardoRossi ?

RiccardoRossi commented:
dof.IsFixed() will tell you if it is fixed or not

On Tue, Mar 3, 2020 at 5:41 PM emiroglu <notifications@github.com> wrote:

> *@emiroglu* commented on this pull request.
> ------------------------------
>
> In
> applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp
> <https://github.com/KratosMultiphysics/Kratos/pull/6434#discussion_r387148794>
> :
>
> >                  for (std::size_t j = 0; j < NumNodeDofs; j++)
>                  {
>                      auto itDof = std::begin(NodeDofs) + j;
> -                    rNodeEigenvectors(i,j) = rEigenvectors(i,(*itDof)->EquationId());
> +                    if ((*itDof)->IsFree()) {
>
> there is another problem in this line:
> For cases with mixed elements (e.g. trusses & beams) on nodes that belong
> only to trusses we also have the ROTATION dofs. These dofs are not in the
> DofSet in the B&S, but they are still there. Meaning that if we loop the
> dofs (which we are doing here) then we also get those "inactive" dofs. They
> return the default equation Id which is 0 and leads to wrong results.
> I don't know to differentiate the "active" from the "inactive" dofs. We
> thought of setting an equation Id of -1 but this does not work bcs it is a
> std::size_t
>
> Maybe someone has an idea?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6434?email_source=notifications&email_token=AB5PWELHOIQMIDOKVKHYGMLRFUXKDA5CNFSM4KXPI5XKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXYWJMA#discussion_r387148794>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWELC2HOFGYUPFZHV2GLRFUXKDANCNFSM4KXPI5XA>
> .
>


-- 


*Riccardo Rossi*

PhD, Civil Engineer


member of the Kratos Team: www.cimne.com/kratos

Associate Professor at Universitat Politècnica de Catalunya, BarcelonaTech
(UPC)

Full Research Professor at International Center for Numerical Methods in
Engineering (CIMNE)


C/ Gran Capità, s/n, Campus Nord UPC, Building C1, First Floor

08034 – Barcelona – Spain – www.cimne.com  -

T.(+34) 93 401 56 96 skype: *rougered4*



<http://www.cimne.com/>

<https://www.facebook.com/cimne> <http://blog.cimne.com/>
<http://vimeo.com/cimne> <http://www.youtube.com/user/CIMNEvideos>
<http://www.linkedin.com/company/cimne> <https://twitter.com/cimne>

Les dades personals contingudes en aquest missatge són tractades amb la
finalitat de mantenir el contacte professional entre CIMNE i voste. Podra
exercir els drets d'accés, rectificació, cancel·lació i oposició,
dirigint-se a cimne@cimne.upc.edu. La utilització de la seva adreça de
correu electronic per part de CIMNE queda subjecte a les disposicions de la
Llei 34/2002, de Serveis de la Societat de la Informació i el Comerç
Electronic.

 Imprimiu aquest missatge, només si és estrictament necessari.
<http://www.cimne.com/>

emiroglu commented:
> dof.IsFixed() will tell you if it is fixed or not

It does but even though the `ROTATION` dofs are normally not fixed for truss elements' nodes, they still exist at the node level since `rotation_dofs=true`. When these dofs are asked for their EquationId they return the default value 0 because they are not assigned anything. So each "inactive" dof that is asked for its value returns whatever in EquationId=0 is.

A fix from the user side can be to fix all the "inactive" dofs in the model, which is very error prone in a complicated model.

RiccardoRossi commented:
why don t you get the dofset from tje strategy?

that is guaranteed to only contain active dofs

On Tue, Mar 3, 2020, 7:25 PM emiroglu <notifications@github.com> wrote:

> dof.IsFixed() will tell you if it is fixed or not
>
> It does but even though the ROTATION dofs are normally not fixed for
> truss elements' nodes, they still exist at the node level since
> rotation_dofs=true. When these dofs are asked for their EquationId they
> return the default value 0 because they are not assigned anything. So each
> "inactive" dof that is asked for its value returns whatever in EquationId=0
> is.
>
> A fix from the user side can be to fix all the "inactive" dofs in the
> model, which is very error prone in a complicated model.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6434?email_source=notifications&email_token=AB5PWEI7MFIMIFHRNG2Y6Z3RFVDQVA5CNFSM4KXPI5XKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENUTFCI#issuecomment-594096777>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEPULERN247L63U3TY3RFVDQVANCNFSM4KXPI5XA>
> .
>

philbucher commented:
the problem is that we need the inactive dofs, not the active ones. 
=> we are looping the nodes and then the dofs of the nodes [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp#L671-L696)
here is the pseudocode:
~~~
for node in Nodes:
    for dof in node.dofs:
        if dof.IsFree and dof in DofSet:
            # do sth
~~~

what we could do is if the current dof is in the DofSet, but this is kinda expensive...
The larger the problem the slower it becomes

Edit updated pseudocode to use dofset

emiroglu commented:
A similar problem will occur while trying to generate POD Basis for `RomApp` because there each node is assumed to have the same number of dofs (i.e. `DISPLACEMENT` and `ROTATION`)

[Conventions for NodalModes.json](https://github.com/KratosMultiphysics/Kratos/issues/6471#issuecomment-592677312) #6471

RiccardoRossi commented:
about looping on "active", free dofs they ARE in the doflist, so i cannot understand what you are asking.

regarding the use of different number of dofs in the rom builder we are working with @Rbravo555  on a solution (i think a prototype is already in place). In the prototype however we assume that the dimension of the ROM_BASIS on each node is the same, even though some of the dofs may be misisng

emiroglu commented:
```
for node in Nodes:
    for dof in node.dofs:
        if dof.IsFree and dof in DofSet:
            # set dof value to result(dof-> EquationId())
        else if !(dof.IsFixed) and dof not in DofSet:
            # set dof value to zero
```
I think the last part was missing from @philbucher 's pseudocode

emiroglu committed:
checking if dof is in DofSet

emiroglu committed:
removing const

emiroglu committed:
checking active dofs

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:
```diff
@@ -678,21 +695,20 @@ class EigensolverStrategy
-                    auto itDof = std::begin(NodeDofs) + j;
-                    rNodeEigenvectors(i,j) = rEigenvectors(i,(*itDof)->EquationId());
+                    const auto itDof = std::begin(NodeDofs) + j;
+                    bool is_active = !(*(r_dof_set.find(**itDof)) == *(r_dof_set.end()));
```
@RiccardoRossi this is what we meant

emiroglu committed:
fixing search for active dofs

philbucher committed:
Merge branch 'master' into structural/improve-bcs-eigensolver

philbucher committed:
update py-exposure

philbucher committed:
using ApplyDirichletConditions of BuilderAndSolver

philbucher committed:
cleanup and improve py-solver

philbucher committed:
correcting json

philbucher committed:
diagonal value CAN be zero!

philbucher committed:
update test

roigcarlo committed:
Dirty fix for the msvc 16.5 std::sort array with kratos array_1d class

philbucher committed:
modifying tests to use block and elim B&S

philbucher committed:
added new test with different dofs

philbucher committed:
cleanup py-EigenSolver

philbucher committed:
finalizign eigensolver-strategy

philbucher committed:
minor cleanup

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:
```diff
@@ -668,33 +679,33 @@ class EigensolverStrategy
-                    auto itDof = std::begin(NodeDofs) + j;
-                    rNodeEigenvectors(i,j) = rEigenvectors(i,(*itDof)->EquationId());
+                    const auto itDof = std::begin(NodeDofs) + j;
+                    bool is_active = !(r_dof_set.find(**itDof) == r_dof_set.end());
```
I checked whether this has an effect on performance
In a Model with 500K Nodes it had no measureable impact

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:
```diff
@@ -394,29 +396,34 @@ class EigensolverStrategy
-        // SPD property
+        // if the size of the Matrix is the same as the size of the dofset, then also the dirichlet-dofs are in the matrix
+        // i.e. a BlockBuilder is used
+        const bool matrix_contains_dirichlet_dofs = SparseSpaceType::Size1(rMassMatrix) == this->pGetBuilderAndSolver()->GetDofSet().size();
```
this is honestly not the cleanest solution, but it is not possible to use `ApplyDirichletConditions` of the B&S, the implementation/requirements are slightly different

philbucher commented:
this is ready, now it should work with Block and Elim B&S as well as when Elements with different DOFs are used
I added a couple of tests for all cases

I have no idea why the CI is failing, for me it works locally :/
Maybe has sth to do with the ordering in the solution.
Does anyone have an idea?
EDIT: nevermind, I am stupid -,-

philbucher committed:
Merge remote-tracking branch 'origin/master' into structural/improve-bcs-eigensolver

philbucher committed:
Merge remote-tracking branch 'origin/unittes/print-failed-index' into structural/improve-bcs-eigensolver

philbucher committed:
adding print

philbucher committed:
introduced small asymmetry to make results deterministic

philbucher committed:
fix skiptest

philbucher committed:
check if MKL is the problem

philbucher commented:
seems like this has the same problem as #5938, again MKL gives a different (but not wrong) result that makes the test fail
Hence I will go ahead with #6573

philbucher commented:
> seems like this has the same problem as #5938, again MKL gives a different (but not wrong) result that makes the test fail
> Hence I will go ahead with #6573

FYI @manuelmessmer

philbucher committed:
Merge remote-tracking branch 'origin/eigensolvers/mkl-optional' into structural/improve-bcs-eigensolver

philbucher committed:
revert CI

philbucher committed:
not using MKL

philbucher committed:
checking abs values

philbucher committed:
lower tolerance

there is already a comment saying this might be too strict, which with the ElimB&S it seems to be
Still very small

philbucher committed:
Merge remote-tracking branch 'origin/master' into structural/improve-bcs-eigensolver

philbucher commented:
this is ready to be reviewed
I checked also a large example with mixed dofs, works too with both Block and Elim B&S

philbucher committed:
Merge branch 'master' into structural/improve-bcs-eigensolver

qaumann commented on applications/StructuralMechanicsApplication/custom_python/add_custom_strategies_to_python.cpp:
```diff
@@ -93,7 +93,18 @@ void  AddCustomStrategiesToPython(pybind11::module& m)
+                py::arg("builder_and_solver"),
+                py::arg("mass_matrix_diagonal_value"),
+                py::arg("stiffness_matrix_diagonal_value"),
+                py::arg("compute_model_decomposition")=false)
```
```suggestion
                py::arg("compute_modal_decomposition")=false)
```

qaumann commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py:
```diff
@@ -73,7 +74,27 @@ def _create_mechanical_solution_strategy(self):
+            if not diag_values.Has("mass_matrix_diagonal_value") or not diag_values.Has("stiffness_matrix_diagonal_value"):
+                err_msg  = 'For the used eigensolver "{}" no defaults for '
+                err_msg += '"mass_matrix_diagonal_value" and "stiffness_matrix_diagonal_value" exist, '
+                err_msg += 'please specify them under "eigensolver_diagonal_values"'
```
I think you a re missing a `.format()`?

qaumann reviewed (dismissed):
Looks good, sorry for blocking. Just two minor things.

philbucher committed:
missing format

philbucher committed:
fix name

philbucher commented:
thx @qaumann for spotting
